### ISSUE TITLE
Emotes incorrectly parsed

### ISSUE THREAD

**Issue by ccKep** [Zibbp/ganymede#413]
It seems parsing emotes from live-chat.json has an off-by-one error which causes Emotes and Text Fragments to be incorrectly created (this doesn't affect every message, I have yet to recognize a pattern though):

![image](https://github.com/Zibbp/ganymede/assets/8322241/f323370d-ff3b-4ebe-aedd-c972dedf0030)

The emotes in question are narilaCONFETTI (that's where the capital I is coming from) and narilaPartyTime2 (that's where the 2 is coming from).

This is the relevant part from the live-chat.json:

```json
    "emotes": [
      {
        "id": "emotesv2_9ffa507158104531b5b82aaff0b51cf4",
        "images": [...],
        "locations": [
          "46-59"
        ],
        "name": "narilaCONFETTI"
      },
      {
        "id": "emotesv2_1a2ac89154474f6197b93195dd141599",
        "images": [...],
        "locations": [
          "61-76"
        ],
        "name": "narilaPartyTime2"
      }
    ],
```

Seeing this, the emote locations match just fine
![image](https://github.com/Zibbp/ganymede/assets/8322241/2f073cb0-2e44-48a3-b30d-a0a21515be0d)

This is the relevant part from chat-convert.json:

```json
"message": {
  "body": "Guten Abend Leute. Schön euch wieder zu sehen narilaCONFETTI narilaPartyTime2",
  "bits_spent": 0,
  "fragments": [
    {
      "text": "Guten Abend Leute. Schön euch wieder zu sehen",
      "emoticon": null,
      "pos1": 0,
      "pos2": 0
    },
    {
      "text": " narilaCONFETT",
      "emoticon": {
        "emoticon_id": "emotesv2_9ffa507158104531b5b82aaff0b51cf4",
        "emoticon_set_id": ""
      },
      "pos1": 46,
      "pos2": 60
    },
    {
      "text": "I",
      "emoticon": null,
      "pos1": 0,
      "pos2": 0
    },
    {
      "text": " narilaPartyTime",
      "emoticon": {
        "emoticon_id": "emotesv2_1a2ac89154474f6197b93195dd141599",
        "emoticon_set_id": ""
      },
      "pos1": 61,
      "pos2": 77
    },
    {
      "text": "2",
      "emoticon": null,
      "pos1": 0,
      "pos2": 0
    }
  ],
}
```

The end positions are off by 1 and the last letter from the emote got put into it's own text fragment. Furthermore there's a blank space at the start of the emote in emote-fragments suggesting that start- and end-offsets are off by 1.

I have no prior experience in Go, but looking at ConvertTwitchLiveChatToVodChat in internal/utils/chat.go Line L263 seems to be the problem in addition to 288-320.

When I parsed emotes in the past it actually helped me to replace them in the message in reverse order (so not to affect later positions and save me the headache from calculating offsets).

eg. replace the last emote first, then the 2nd to last etc. - this means you can always just use the original offsets.

**Comment by Zibbp** [comment]
Can you upload the full unconverted `live-chat.json` file, so I can reproduce this?

**Comment by ccKep** [comment]
Ofcourse, here you go.

I zipped it since Github only allows 25MB Uploads:
[50977473517-live-chat.zip](https://github.com/Zibbp/ganymede/files/15160049/50977473517-live-chat.zip)

**Comment by Zibbp** [comment]
What appears to be happening is special characters, in this case umlauts, are not being counted in [chat-downloader](https://github.com/xenova/chat-downloader) when the positions of emotes are calculated. The raw live message shows the unicode for the special character
```json
"message": "Guten Abend Leute. Sch\u00f6n euch wieder zu sehen narilaCONFETTI narilaPartyTime2",
```
This results in messages containing these having their emotes positions off by N.

I can't simply use `strings.Index` to retrieve the real position of the emote as there might be multiple of the same emote in the message. I also can't utilize the starting position to filter repeats of the emote as that is incorrect in this context.

I can probably make a function to iterate over the string and get the start and end positions of N occurrences of the emote in the string.

**Comment by Zibbp** [comment]
I've merged https://github.com/Zibbp/ganymede/pull/425, which should be available in the `:main` image tag. If the extracted emote from the given start and end positions does not match the emote's name, it now attempts to extract the emote positions from the message. I tested against the chat you supplied and it is properly splitting the emotes now.
```json
      "message": {
        "body": "Guten Abend Leute. Schön euch wieder zu sehen narilaCONFETTI narilaPartyTime2",
        "bits_spent": 0,
        "fragments": [
          {
            "text": "Guten Abend Leute. Schön euch wieder zu sehen ",
            "emoticon": null,
            "pos1": 0,
            "pos2": 0
          },
          {
            "text": "narilaCONFETTI",
            "emoticon": {
              "emoticon_id": "emotesv2_9ffa507158104531b5b82aaff0b51cf4",
              "emoticon_set_id": ""
            },
            "pos1": 47,
            "pos2": 61
          },
          {
            "text": " ",
            "emoticon": null,
            "pos1": 0,
            "pos2": 0
          },
          {
            "text": "narilaPartyTime2",
            "emoticon": {
              "emoticon_id": "emotesv2_1a2ac89154474f6197b93195dd141599",
              "emoticon_set_id": ""
            },
            "pos1": 62,
            "pos2": 78
          },
          {
            "text": "",
            "emoticon": null,
            "pos1": 0,
            "pos2": 0
          }
        ],
```

**Comment by ccKep** [comment]
I'll test this as soon as possible. I'm still running ganymede 1.4.3 as my main instance (although I verified the emote behaviour on v2 before filing the report ofcourse). I'll probably upgrade to v2 soon, thanks alot!

**Comment by ccKep** [comment]
Just updated to v2, is there any way to re-convert the chat files so I can test the fix retroactively?

**Comment by Zibbp** [comment]
Coming from a non 2.X version, no. You can easily test this fix by archiving a stream that has a lot of unicode messages, like the umlaut in the original chat.

**Comment by ccKep** [comment]
Just recorded the first stream, unfortunately it did not process as expected:

![image](https://github.com/Zibbp/ganymede/assets/8322241/ff8de175-3b67-47a7-a2b4-7451b01c35d9)

![image](https://github.com/Zibbp/ganymede/assets/8322241/7e960d0a-2e01-4277-be69-802e7773792d)

Looking at the temporal web UI:
![image](https://github.com/Zibbp/ganymede/assets/8322241/59900734-43b2-4d96-a61a-7890809daba6)

`{
  "message": "runtime error: slice bounds out of range [13:0]",
  "source": "GoSDK",
  "stackTrace": "activity for archive [panic]:
github.com/zibbp/ganymede/internal/utils.ConvertTwitchLiveChatToTDLChat({0xc0020c0140, 0x44}, {0xc001f30f18, 0x7}, {0xc002082720, 0x24}, {0xc001f30ee0, 0xb}, 0x4965480, {0xbfaa68, ...}, ...)
	/app/internal/utils/tdl.go:250 +0x117b
github.com/zibbp/ganymede/internal/activities.ConvertTwitchLiveChat({0x1663988, 0xc000902780}, {{0xc001f30ec0, 0xb}, {0xc001f30ea8, 0x4}, {0xc001f30ed0, 0x6}, {0xc001f30ed8, 0x4}, ...})
	/app/internal/activities/video.go:835 +0x11bc
reflect.Value.call({0x1267aa0?, 0x14e1f18?, 0x30?}, {0x145a51e, 0x4}, {0xc0009027b0, 0x2, 0xc0009027b0?})
	/usr/local/go/src/reflect/value.go:596 +0xca6
reflect.Value.Call({0x1267aa0?, 0x14e1f18?, 0x2a?}, {0xc0009027b0?, 0x30?, 0x7fd977475f70?})
	/usr/local/go/src/reflect/value.go:380 +0xb9
go.temporal.io/sdk/internal.executeFunction({0x1267aa0, 0x14e1f18}, {0xc001f32dc0, 0x2, 0x126a6e0?})
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker.go:1858 +0x26b
go.temporal.io/sdk/internal.executeFunctionWithContext({0x1663988, 0xc000902780}, {0x1267aa0, 0x14e1f18}, {0xc00065a070, 0x1, 0x1})
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker.go:1842 +0x1ae
go.temporal.io/sdk/internal.(*activityEnvironmentInterceptor).ExecuteActivity(0xc002008c40, {0x1663a30?, 0xc0003b3b90?}, 0xc001004540)
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_activity.go:366 +0x67
go.temporal.io/sdk/internal.(*activityExecutor).ExecuteWithActualArgs(0xc0003d0f90, {0x1663a30, 0xc0003b3b90}, {0xc00065a070, 0x1, 0x1})
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker.go:859 +0x148
go.temporal.io/sdk/internal.(*activityExecutor).Execute(0xc0003d0f90, {0x1663a30, 0xc0003b3b90}, 0xc001e78900)
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker.go:845 +0x139
go.temporal.io/sdk/internal.(*activityTaskHandlerImpl).Execute(0xc00080c0e0, {0x145ddb1, 0x7}, 0xc0006b7ad0)
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_task_handlers.go:2117 +0x924
go.temporal.io/sdk/internal.(*activityTaskPoller).ProcessTask(0xc000373290, {0x11ea5c0, 0xc000922088})
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_task_pollers.go:1001 +0x283
go.temporal.io/sdk/internal.(*baseWorker).processTask(0xc0003c5080, {0x11ea680, 0xc0003a18d0})
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker_base.go:518 +0x153
go.temporal.io/sdk/internal.(*baseWorker).processTaskAsync.func1()
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker_base.go:369 +0x45
created by go.temporal.io/sdk/internal.(*baseWorker).processTaskAsync in goroutine 87
	/go/pkg/mod/go.temporal.io/sdk@v1.26.1/internal/internal_worker_base.go:365 +0xa5",
  "applicationFailureInfo": {
    "type": "PanicError"
  }
}`

**Comment by ccKep** [comment]
https://github.com/Zibbp/ganymede/blob/main/internal/utils/tdl.go#L199C1-L200C99
`						// attempt to get emote position in comment message
						pos1, pos2, found := findSubstringPositions(liveComment.Message, liveCommentEmote.Name, i+1)`

simply using i+1 (basically the i'th emote in the string when starting from 1) seems flawed here aswell since not every emote gets into this branch of the code, so you could be searching for the 7th emote overall but it there might be only 2 occurences of a problematic emote due to position?

**Comment by Zibbp** [comment]
Can you upload the live chat json file of the failed archive?

> simply using i+1 (basically the i'th emote in the string when starting from 1) seems flawed here aswell since not every emote gets into this branch of the code, so you could be searching for the 7th emote overall but it there might be only 2 occurences of a problematic emote due to position?

I don't follow. The `findSubstringPositions` is running within a `for` loop for each emote position. This means that `i+1` will always match up with the `Nth` position of a emote.

Take the following message from your first chat as an example.

```json
    "message": "narilaHERZGIF Als DANKESCH\u00d6N f\u00fcr meine Ko-fi's, k\u00f6nnen Abonnenten Zugriff auf 344 VODs mit 75 Tagen, 7 Stunden, 54 Minuten und 50 Sekunden Content bekommen. 24/7 Stream, wann immer du willst? Klick mich: https://vods.narilah.de/vods narilaHERZGIF",
```
 The offset emote positions do not start until the first unicode character. So the first `narilaHERZGIF` is extracted using the provided positions. The seconds `narilaHERZGIF` is extracted using the `findSubstringPositions` function. `i+1` works here as this is the second `narilaHERZGIF` emote of the message. Resulting output is

```json
        "fragments": [
          {
            "text": "",
            "emoticon": null,
            "pos1": 0,
            "pos2": 0
          },
          {
            "text": "narilaHERZGIF",
            "emoticon": {
              "emoticon_id": "emotesv2_dde4b5bfcdd141ea8d80ec7f0852de1b",
              "emoticon_set_id": ""
            },
            "pos1": 0,
            "pos2": 13
          },
          {
            "text": " Als DANKESCHÖN für meine Ko-fi's, können Abonnenten Zugriff auf 344 VODs mit 75 Tagen, 7 Stunden, 54 Minuten und 50 Sekunden Content bekommen. 24/7 Stream, wann immer du willst? Klick mich: https://vods.narilah.de/vods ",
            "emoticon": null,
            "pos1": 0,
            "pos2": 0
          },
          {
            "text": "narilaHERZGIF",
            "emoticon": {
              "emoticon_id": "emotesv2_dde4b5bfcdd141ea8d80ec7f0852de1b",
              "emoticon_set_id": ""
            },
            "pos1": 236,
            "pos2": 249
          },
          {
            "text": "",
            "emoticon": null,
            "pos1": 0,
            "pos2": 0
          }
        ],
```

**Comment by ccKep** [comment]
omg, I'm so sorry - I misread and was under the impression `i` iterates over all emotes and not per-emote. In that case you can ignore that comment ofcourse!

Here you have the live chat in question: 
[42336355048-live-chat.zip](https://github.com/Zibbp/ganymede/files/15319678/42336355048-live-chat.zip)

**Comment by Zibbp** [comment]
I've pushed some more changes to https://github.com/Zibbp/ganymede/pull/433 which should handle this issue with the latest chat file. The new issue was because the `<3` emote's `<` is unicode in the live chat file. The function to parse emote positions couldn't handle unicode. 

I don't archive any streams that have non-english characters so it's hard for me to fully test these changes. If you want to test you can pull `:main` image and try it out.

**Comment by ccKep** [comment]
Seems to be working, thanks alot!
